fsthttp/request.go

```diff
@@ -399,40 +423,308 @@ func (req *Request) Send(ctx context.Context, backend string) (*Response, error)
+	req          *Request
+}
+
+func (req *Request) sendAsyncForCaching(ctx context.Context, cacheHandle *fastly.HTTPCacheHandle, backend string) (*pendingBackendRequestForCaching, error) {
+
+	reqh, err := fastly.HTTPCacheGetSuggestedBackendRequest(cacheHandle)
+	if err != nil {
+		return nil, fmt.Errorf("get suggested backend request: %w", err)
+	}
+
+	suggReq, err := newRequestFromHandle(reqh, req.Body, req.Header, req.CacheOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	if suggReq.CacheOptions.BeforeSend != nil {
+		if err := suggReq.CacheOptions.BeforeSend(suggReq); err != nil {
+			// TODO(dgryski): sentinel ErrReject ?
+			return nil, err
+		}
+	}
+
+	// If BeforeSend calls SetBody, abi.body will be updated for the new Body
+	finalCacheOptions := suggReq.CacheOptions
+	suggReq.setABIRequestOptions()
+
+	// copy body
+	var errc chan error = make(chan error, 3)
+	suggReq.copyBody(errc)
+
+	abiPending, err := suggReq.sendAsyncWithoutCaching(ctx, backend)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pendingBackendRequestForCaching{
+		cacheHandle:  cacheHandle,
+		req:          suggReq,
+		pending:      abiPending,
+		afterSend:    req.CacheOptions.AfterSend,
+		cacheOptions: finalCacheOptions,
+	}, nil
+}
+
+func (req *Request) sendAsyncWithoutCaching(_ context.Context, backend string) (*fastly.PendingRequest, error) {
+	abiPending, err := req.abi.req.SendAsyncV2(req.abi.body, backend, false)
+
+	if err != nil {
+		return nil, fmt.Errorf("send async: %w", err)
+	}
+
+	return abiPending, nil
+}
+
+func (req *Request) sendWithoutCaching(backend string) (*fastly.HTTPResponse, *fastly.HTTPBody, error) {
+	abiResp, abiBody, err := req.abi.req.SendV3(req.abi.body, backend)
+	if err != nil {
+		return nil, nil, fmt.Errorf("send v3: %w", err)
+	}
+	return abiResp, abiBody, nil
+}
+
+func (req *Request) mustUseHostCaching() (bool, error) {
+	ok, err := fastly.HTTPCacheIsRequestCacheable(req.abi.req)
+	return !ok, err
+}
+
+func (req *Request) shouldUseGuestCache() (bool, error) {
+	// disabled via buildtags or unsupported hostcalls
+	if !useGuestCaching {
+		if req.CacheOptions.mustUseGuestCaching() {
+			return false, errors.New(fastly.FastlyStatusUnsupported.String())
+		}
+
+		return false, nil
+	}
+
+	if req.CacheOptions.Pass {
+		// skip cache
+		return false, nil
+	}
+
+	if req.Method == "PURGE" {
+		// no caching for PURGE
+		return false, nil
+	}
+
+	mustUseHostCaching, err := req.mustUseHostCaching()
+	// check for hostcall unsupported error
+	if err != nil {
+		if status, ok := fastly.IsFastlyError(err); ok && status == fastly.FastlyStatusUnsupported {
+			// disable for future calls
+			useGuestCaching = false
+			return false, nil
+		}
+		return false, err
+	}
+
+	if mustUseHostCaching {
+		if req.CacheOptions.mustUseGuestCaching() {
+			// oops, before/after hooks need guest caching.
+			// TODO(dgryski): custom fsthttp error here instead?
+			return false, errors.New(fastly.FastlyStatusUnsupported.String())
```

Comment:
Seems like a good idea to differentiate, otherwise this could be confusing to debug.

Reply:
5563b3f